crates/bevy_math/src/curve/adaptors.rs

```diff
@@ -0,0 +1,530 @@
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(Reflect),
+    reflect(from_reflect = false)
+)]
+pub struct LinearReparamCurve<T, C> {
+    /// Invariants: The domain of this curve must always be bounded.
+    pub(crate) base: C,
+    /// Invariants: This interval must always be bounded.
+    pub(crate) new_domain: Interval,
+    #[cfg_attr(feature = "bevy_reflect", reflect(ignore))]
+    pub(crate) _phantom: PhantomData<T>,
+}
+
+impl<T, C> Curve<T> for LinearReparamCurve<T, C>
+where
+    C: Curve<T>,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        self.new_domain
+    }
+
+    #[inline]
+    fn sample_unchecked(&self, t: f32) -> T {
+        // The invariants imply this unwrap always succeeds.
+        let f = self.new_domain.linear_map_to(self.base.domain()).unwrap();
+        self.base.sample_unchecked(f(t))
+    }
+}
+
+/// A curve that has been reparametrized by another curve, using that curve to transform the
+/// sample times before sampling. Curves of this type are produced by [`Curve::reparametrize_by_curve`].
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
+pub struct CurveReparamCurve<T, C, D> {
+    pub(crate) base: C,
+    pub(crate) reparam_curve: D,
+    #[cfg_attr(feature = "bevy_reflect", reflect(ignore))]
+    pub(crate) _phantom: PhantomData<T>,
+}
+
+impl<T, C, D> Curve<T> for CurveReparamCurve<T, C, D>
+where
+    C: Curve<T>,
+    D: Curve<f32>,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        self.reparam_curve.domain()
+    }
+
+    #[inline]
+    fn sample_unchecked(&self, t: f32) -> T {
+        let sample_time = self.reparam_curve.sample_unchecked(t);
+        self.base.sample_unchecked(sample_time)
+    }
+}
+
+/// A curve that is the graph of another curve over its parameter space. Curves of this type are
+/// produced by [`Curve::graph`].
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
+pub struct GraphCurve<T, C> {
+    pub(crate) base: C,
+    #[cfg_attr(feature = "bevy_reflect", reflect(ignore))]
+    pub(crate) _phantom: PhantomData<T>,
+}
+
+impl<T, C> Curve<(f32, T)> for GraphCurve<T, C>
+where
+    C: Curve<T>,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        self.base.domain()
+    }
+
+    #[inline]
+    fn sample_unchecked(&self, t: f32) -> (f32, T) {
+        (t, self.base.sample_unchecked(t))
+    }
+}
+
+/// A curve that combines the output data from two constituent curves into a tuple output. Curves
+/// of this type are produced by [`Curve::zip`].
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect))]
+pub struct ProductCurve<S, T, C, D> {
```

Comment:
The name `ProductCurve` might be confused for the numerical product of the data from the two curves (at least when I read this type name initially, that's what I thought it was).

Since the fn is already `Curve::zip`, could this just be called `ZipCurve`?

Reply:
Good idea. Done!